Pattern acceleration

elliVM opened the pull request:
Pattern acceleration feature to activate bloom filtering on set regex pattern. Goal is to limit bloom filtering to certain patterns like UUID.

Flow:
- Walker walks XML and for each search term regex checks tokenized term against tables in bloomdb excluding the filtertype table
- Collect list of tables with a pattern match record from SQL metadata
- For each table a temp table is created
- For each temp table generate condition that compares temp table filter against parent table filter using bloommatch UDF
- Combine conditions from each table, if all filters were null then also true
- Combine conditions from each search term
- Left join tables to NestedTopNQuery and use the final combined condition in query

Notes: Tokenizer max token count is set to 0 to get only major tokens since that is what dpf_03 currently uses to tokenize the bloom filter tables

eemhu commented on src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:
```diff
@@ -0,0 +1,236 @@
+        this.ctx = ctx;
+        this.parentTable = parentTable;
+        this.bloomTermId = bloomTermId;
+        this.tokenSet = tokenSet;
```
Perhaps this could be named as searchTokens or similar? to clarify which tokens are used.

elliVM replied:
renamed to searchTermTokenSet

eemhu commented on src/main/java/com/teragrep/pth_06/planner/walker/ConditionWalker.java:
```diff
@@ -68,27 +73,37 @@ public class ConditionWalker extends XmlWalker<Condition> {
     // Default query is full
     private boolean streamQuery = false;
     private final DSLContext ctx;
```
is it possible to make these final?

elliVM replied:
made final also ValidElement class is final now

eemhu replied:
what about `streamQuery` and `bloomTermId`?

elliVM replied:
making `streamQuery` final would be possible using separate walkers for stream and journal conditions but that would need some refactoring of testing that I would do in another PR.

elliVM replied:
`bloomTermId` would also a bigger refactor to make work with walker traverse on a different PR

elliVM replied:
will make issues for  both

eemhu commented on src/test/java/com/teragrep/pth_06/planner/walker/conditions/ElementConditionTest.java:
```diff
@@ -84,20 +83,6 @@ void testStreamTags() {
 
-    @Test
-    void testIndexStatement() {
-        Element element = document.createElement("indexstatement");
```
what happened to this test?

elliVM replied:
I moved indexstatement testing to IndexStatementConditionTests but I added a testIndexStatementWithBadConnection to ElementConditionTest that checks for DataAccessException with a bad connection. Also added same check for IndexStatementConditionTest.

elliVM commented on src/main/java/com/teragrep/pth_06/planner/walker/conditions/ElementCondition.java:
```diff
@@ -119,18 +115,37 @@ public Condition condition() {
     }
 
+    public boolean isIndexStatement() {
+        String tag = element.tag();
```
renamed to isBloomSearchCondition() which is a better description of the function also added tests for the function

kortemik commented on src/main/java/com/teragrep/pth_06/planner/StreamDBClient.java:
```diff
@@ -420,15 +343,17 @@ private Table<Record> getTableStatement(Condition journaldbCondition, Date day)
-                        .leftJoin(Bloomdb.BLOOMDB.FILTER_EXPECTED_2500000_FPP_005)
-                        .on(JOURNALDB.LOGFILE.ID.eq(Bloomdb.BLOOMDB.FILTER_EXPECTED_2500000_FPP_005.PARTITION_ID));
-
+                Set<Table<?>> tables = walker.patternMatchTables();
```
please elaborate what happens here

elliVM replied:
All bloom filter tables that had a pattern match during walker traverse are left joined to the query. Dynamic number of bloom filter tables can be joined depending on number of pattern matches while keeping one to one relationship between `journaldb.logfile.id` and the bloom filter tables

kortemik commented on src/main/java/com/teragrep/pth_06/planner/walker/conditions/ElementCondition.java:
```diff
@@ -119,18 +115,39 @@ public Condition condition() {
+                && config.bloomEnabled();
+    }
+
+    public Set<Table<?>> patternMatchTables() {
```
why this is present in both, ElementCondition and the walker:
https://github.com/teragrep/pth_06/pull/87/files#diff-58946539ef28e538e946fbe84b99864bf3f4a912e9dfab8b94f4a206c21bbbf0R103

elliVM replied:
ElementCondition finds pattern match tables from a single element, while walker keeps list of pattern match tables from all traversed elements

elliVM commented:
## Refactoring 

New classes:
| Class | Responsibility |
| --- | --- |
| PatternMatchTables | Finds bloomdb Tables that match a pattern condition |
|CategoryTableImpl |  Temp table from a bloom filter table that can return a `CategoryTableCondition` |
|Created | Decorator that creates `CategoryTable` to database |
|WithFilterTypes | Decorator that inserts filters of category table, gets different types from the table that `CategoryTable` was created from. Search term tokens are inserted into category table filters|
|TableFilters | Different filter types of a `CategoryTable` |
|TableFilterTypesFromMetadata | Fetches different filter types of a table from metadata |
|CategoryTableCondition | Condition that compares category tables filter bytes against correct boom filter table filter bytes, selects the same size and bloom term id|
|PatternMatchCondition|Condition that  check if any of given tokens match with `bloomdb.filtertype.pattern `|

Other changes:

- Many method naming changes
- Added interfaces `CategoryTable`, `TableRecords`, `BloomQueryCondition`
- Added missing `withoutFilters` option and implementation to `IndexStatementCondition`
- Equality methods for all new classes
- Tests for all new classes

elliVM commented on src/main/java/com/teragrep/pth_06/config/ConditionConfig.java:
```diff
@@ -52,20 +52,15 @@ public final class ConditionConfig {
+        this(ctx, streamQuery, false);
     }
 
-    public ConditionConfig(DSLContext ctx, boolean streamQuery, boolean bloomEnabled, boolean withoutFilters) {
```
Option was not used in github repo. Added option with implementation and tests

elliVM commented on src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:
```diff
@@ -0,0 +1,236 @@
+     * Inserts a filter for each filtertype inside parent table records. Filter is filled with search term token set,
+     * filter size is selected using parent table filtertype expected and fpp values.
+     */
+    private void insertFilters() {
```
Method was poorly named. After refactoring  `TableFilters` is responsible of category tables filters and inserting all filter types. class `TableFilterTypesFromMetadata` is used to fetch the tables filter types from metadata

elliVM commented on src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:
```diff
@@ -0,0 +1,236 @@
+        }
+    }
+
+    private void insertFilterFromRecord(
```
refactored into method  `filterBytesFromRecord` in class `TableFilters` that creates a bloom filter with correct size and tokens.

elliVM commented on src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    /**
+     * Generates a condition that returns true if this temp tables search term tokens might be contained in the parent
```
Parent table was the bloom filter table that the temp table was created to match, renamed as `originTable` in new class `CategoryTableImpl`

elliVM commented on src/main/java/com/teragrep/pth_06/planner/PatternMatch.java:
```diff
@@ -0,0 +1,137 @@
+
+import static com.teragrep.pth_06.jooq.generated.bloomdb.Bloomdb.BLOOMDB;
+
+public class PatternMatch {
```
Responsibility was to get bloom filter tables that had record with a filtertype pattern that matched with any of the search term tokens. Refactored into two classes `PatternMatchTables` thet gets the tables from metadata and `PatternMatchCondition` that generates the condition used to do the pattern matching

eemhu commented on src/main/java/com/teragrep/pth_06/planner/Created.java:
```diff
@@ -0,0 +1,76 @@
+
+/**
+ * Decorator that creates category table
+ */
```
The purpose of this object is not super clear to me, why not just call create() on the origin categoryTable?

elliVM replied:
Removed decorators in refactoring and just use `CategoryTable` interface methods `create()` and `insertFilters()`.

eemhu commented on src/test/java/com/teragrep/pth_06/planner/CategoryTableImplTest.java:
```diff
@@ -0,0 +1,312 @@
+        TokenizedValue val = new TokenizedValue("test");
+        CategoryTableImpl tempTable = new CategoryTableImpl(ctx, table, 0L, val);
+        QueryCondition tableCond = tempTable.bloommatchCondition();
+        QueryCondition expectedCond = new CategoryTableCondition(table, 0L);
```
Would it be better to create expected queryCondition "manually" instead of relying on another class?
If both classes make the same wrong result the test would pass

elliVM replied:
Refactored to compare queryCondition using condition.toString() like in other unit tests

eemhu commented on src/test/java/com/teragrep/pth_06/planner/PatternMatchTablesTest.java:
```diff
@@ -0,0 +1,221 @@
+    final List<String> patternList = new ArrayList<>(
+            Arrays
+                    .asList(
+                            "(\\b25[0-5]|\\b2[0-4][0-9]|\\b[01]?[0-9][0-9]?)(\\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}",
```
I think these patterns could use some more information on what they match

elliVM replied:
created as String objects first and added comments for clarity

eemhu commented on src/test/java/com/teragrep/pth_06/planner/TableFilterTypesFromMetadataResultTest.java:
```diff
@@ -0,0 +1,235 @@
+        Assertions.assertNotEquals(result1, result2);
+        Assertions.assertNotEquals(result2, result1);
+        Assertions.assertNotEquals(result1, result3);
+        Assertions.assertNotEquals(result1, null);
```
why equality tested with null?

elliVM replied:
Cleaned up needless not equals checks against null from tests

eemhu commented on src/test/java/com/teragrep/pth_06/planner/TokenizedValueTest.java:
```diff
@@ -0,0 +1,88 @@
+    void testTokenization() {
+        TokenizedValue result = new TokenizedValue("test.nest");
+        Set<String> tokens = result.tokens().stream().map(Token::toString).collect(Collectors.toSet());
+        Assertions.assertEquals("test.nest", result.value);
```
assert tokens set size, now it could also contain other tokens and still pass

elliVM replied:
added assertion to check for  size

eemhu commented on src/test/java/com/teragrep/pth_06/planner/walker/conditions/CategoryTableConditionTest.java:
```diff
@@ -0,0 +1,228 @@
+        CategoryTableCondition cond1 = new CategoryTableCondition(target1, 1L);
+        CategoryTableCondition cond2 = new CategoryTableCondition(target1, 1L);
+        Assertions.assertEquals(cond1, cond2);
+        Assertions.assertEquals(cond2, cond1);
```
I don't think it is necessary to check cond1==cond2 and then cond2==cond1? spotted this on other tests as well.

elliVM replied:
Cleaned up assertions in equality tests, removed double equality checks and null not equals checks

eemhu commented on src/main/java/com/teragrep/pth_06/planner/walker/conditions/IndexStatementCondition.java:
```diff
@@ -97,9 +97,9 @@ public Condition condition() {
             for (final Table<?> table : tableSet) {
                 // create category temp table for this pattern match table
-                final CategoryTable categoryTable = new Created(
-                        new WithFilterTypes(new CategoryTableImpl(config, table, tokenizedValue))
```
Now that I think about it, perhaps the older way was better to reduce temporal coupling?

kortemik replied:
yes, just with better naming.

eemhu commented on src/test/java/com/teragrep/pth_06/planner/CategoryTableImplTest.java:
```diff
@@ -173,11 +169,24 @@ public void testCreation() {
+                .get(0);
+
+        CategoryTable categoryTable = new CategoryTableImpl(ctx, table, 0L, "192.168.1.1");
+        Assertions.assertDoesNotThrow(categoryTable::create);
```
could this use assertion that the filters are inserted?

elliVM replied:
Yes good spot I updated test to check that only pattern matching tokens are inserted. Changed TokenizedValue to find minor tokens since for example search term `ip=192.168.1.1` did not get token `192.168.1.1`.

elliVM commented on src/main/java/com/teragrep/pth_06/planner/walker/conditions/IndexStatementCondition.java:
```diff
@@ -82,34 +82,28 @@ public Condition condition() {
+                combinedTableCondition = combinedTableCondition.or(tableCondition.condition());
+                combinedNullFilterCondition = combinedNullFilterCondition.and(nullFilterCondition);
             }
+            newCondition = config.withoutFilters() ? combinedNullFilterCondition : combinedTableCondition
```
Changed to if-statement

elliVM commented:
Enabled token entanglement in tokenizer
